### PR TITLE
Newfeature/issue 6 create header component

### LINKED ISSUE #6: create a Header Component
## Type of Change
- [x] New Feature Request
- [ ] Improvement / Suggestion
- [ ] Bug
- [ ] Other (please clarify below)

## Summary

New Feature Request / Improvement:
- Description: A header to show PVD Geeks branding / logo and social network sites
- Designs: See Application Design link above ^
- Dependencies: Designs / Application Design
- Acceptance Criteria:  Based on the Application Design [here](https://github.com/ProvidenceGeeks/website-docs/wiki/Application-Design#header)

### PR DESCRIPTION
**Pull request by Peter-holyoke**
## Related Issue

resolves #6 and #16 

## Summary of Changes

Added a header component. Completely finished for the desktop version.

### CODE REVIEW

File: src/components/header/header.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+export default class Hello extends React.Component {
+  constructor() {
+    super();
+  }
+
+  render() {
+    return (
+      <div className="row-fluid">
+        <div id="header" className="col-xs-12">
```
Comment (ChrisPlease) [inline src/components/header/header.js]: I'd avoid the nested `.col`'s

File: src/components/header/header.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+          <div className="col-xs-2 logo">
+            <Logo />
+          </div>
+
+          <div className="col-xs-7">&nbsp;</div>
```
Comment (ChrisPlease) [inline src/components/header/header.js]: This seems kinda hacky. Instead, you can give `.col-xs-3` another `offset` class. I think it'd be `.col-xs-offset-7` and drop this empty div. Also, bootstrap 3 still has `pull-right` available.

File: src/components/header/header.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+          <div className="social col-xs-3">
```
Comment (ChrisPlease) [inline src/components/header/header.js]: fwiw, the social icons break out of a container this small. You might want to give it a bigger container for smaller viewports, and adjust on larger viewports: 

image

File: src/components/header/header.scss [hunk 4]
```diff
@@ -0,0 +1,60 @@
+$primary-color: rgb(27, 41, 48);
+$logo-color: rgb(246, 205, 77);
+$border-color: rgb(150, 150, 150);
+$social-icon-size: 4rem;
+
+#header {
+  background: $primary-color;
+  display: flex;
+  height: 10rem;
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: Is there any reason for a fixed height here? If it's a matter of containing the size of elements, I'd propose setting a `width` on the element and letting height be it's natural height.

File: src/components/header/header.scss [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  align-items: center;
+  justify-content: flex-end;
+}
+
+.social div {
+  margin-right: $social-icon-size/1.5;
+}
+
+.fa.fa-facebook {
+  color: #fff;
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: IMHO, I'd store all colors as variables. If by chance you decide true white is _too_ white, it's so much easier to update the variable once. Same with social media colors.

File: src/components/header/header.scss [hunk 6]
```diff
@@ -0,0 +1,60 @@
+$primary-color: rgb(27, 41, 48);
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: This is really a matter of preference, but I find it's easier to store all variables in a partial, then import the partial. No need to hunt down the specific variable, and it allows you to handle globals easier.

File: src/components/header/header.scss [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+.logo {
+  margin: 1rem 0 1rem 0;
+}
+
+.logo path {
+  fill: $logo-color;
+}
+
+.logo svg {
+  padding: .75rem 0 .75rem 0;
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: nitpick: you can get this with just `padding: .75rem 0`

File: src/components/bootstrap/bootstrap.scss [hunk 8]
```diff
@@ -0,0 +1,5 @@
+@import url('~bootstrap/dist/css/bootstrap.css');
+
+.container-fluid {
+  padding: 0;
```
Comment (ChrisPlease) [inline src/components/bootstrap/bootstrap.scss]: Why did you need to overwrite bootstraps default style here?

Edit: since you're using `.container-fluid`, you can use `.row` instead of `.row-fluid`. `.row` contains a negative margin that will offset `.container-fluid`'s extra padding, so you don't need to overwrite anything here.

File: src/components/bootstrap/bootstrap.scss [hunk 9]
```diff
@@ -0,0 +1,5 @@
+@import url('~bootstrap/dist/css/bootstrap.css');
```
Comment (ChrisPlease) [inline src/components/bootstrap/bootstrap.scss]: This works, but alternatively, with webpack you should be able to import this directly into the `bootstrap` component:
`import 'bootstrap/dist/css/bootstrap.css';`

File: src/components/header/header.js (hunk 3 above)
Comment (ChrisPlease) [inline src/components/header/header.js]: Also a tiny nitpick, but semantically this is an unordered list:

```
<div className="col-xs-3">
  <ul className="social list-inline">
    <li>...</li>
    <li>...</li>
    <li>...</li>
  </ul>
</div>
```
Makes more sense, semantically.

File: src/components/header/header.scss [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+.slack {
+  width: $social-icon-size;
+}
+
+.circle {
+  background-color: #333;
+  border-radius: $social-icon-size;
+  color: #fff;
+  display: inline-block;
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: `inline-block` isn't really necessary here, it's getting inherited from `.fa` anyway.

File: src/components/header/header.scss [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+
+#header {
+  background: $primary-color;
+  display: flex;
+  height: 10rem;
+  color: #fff;
+}
+
+.logo {
+  margin: 1rem 0 1rem 0;
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: nitpick: same as `margin: 1rem 0`

File: src/components/header/header.js (hunk 1 above)
Comment (ChrisPlease) [inline src/components/header/header.js]: prefer classes over id's

File: src/components/header/header.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+export default class Hello extends React.Component {
+  constructor() {
+    super();
+  }
+
+  render() {
+    return (
+      <div className="row-fluid">
+        <div id="header" className="col-xs-12">
+          <div className="col-xs-2 logo">
```
Comment (ChrisPlease) [inline src/components/header/header.js]: Just as a matter of simplifying things: any div that's acting as a grid piece to bootstrap should _only_ be doing that. I'd recommend nesting this if you need extra styles for `.logo`. Keeping track of overwriting styles gets messy, and you definitely want to avoid overwriting bootstrap's grid styles.

File: src/components/header/header.spec.js [hunk 13]
```diff
@@ -0,0 +1,16 @@
+import * as React from 'react';
+import raf from '../../../test/shims/react-raf-shim';
+import { shallow, configure } from 'enzyme';
+import Adapter from 'enzyme-adapter-react-16';
+
+import Hello from './header';
```
Comment (ChrisPlease) [inline src/components/header/header.spec.js]: nitpick: bad naming here. This is the `Header` component

File: src/components/logo/logo.js [hunk 14]
```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+
+export default class Logo extends React.Component {
+  constructor() {
```
Comment (ChrisPlease) [inline src/components/logo/logo.js]: Do you get an error if you take out the `constructor` here? This seems unnecessary. Also, fwiw, this could just be a functional component instead of a class.

File: src/components/header/header.scss [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+  display: flex;
+  align-items: center;
+  justify-content: flex-end;
+}
+
+.social div {
+  margin-right: $social-icon-size/1.5;
+}
+
+.fa.fa-facebook {
```
Comment (ChrisPlease) [inline src/components/header/header.scss]: Normally, I'd advise about overnesting, but might make things simpler here:

```
.fa {
  color: #fff
  &.fa-facebook {
    background: rgba(...)
  }

  &.fa-twitter {
    background: rgba(...);
  }
}
```

File: src/components/bootstrap/bootstrap.scss (hunk 9 above)
Comment (thescientist13) [reply src/components/bootstrap/bootstrap.scss]: neat!

**Comment by thescientist13** [comment]
@ChrisPlease 
Regarding this [comment](https://github.com/ProvidenceGeeks/website-frontend/pull/13#discussion_r145181476)
> This works, but alternatively, with webpack you should be able to import this directly into the bootstrap component:
`import 'bootstrap/dist/css/bootstrap.css';`

What is the value vs the current approach?  The current decisions aims to allow reusability of _bootstrap.css_ in the following ways:
1.  Other components can just pull in _component/bootstrap/bootstrap.css_ and have access to whatever globals and mixins it comes with (like from vendor files).  By putting it in the JS, you lose that ability, right?  So If we have shared vendor variables for example, how would we use them in the app?
1.  All vendor CSS gets managed via a CSS / SASS / file, which seems a bit more intuitive.

File: src/components/header/header.js [hunk 16]
```diff
@@ -0,0 +1,33 @@
+import * as React from 'react';
+import { Icon } from 'react-fa';
+import './header.scss';
+import Logo from '../../components/logo/logo';
+import SlackLogo from '../../images/Slack-Monochrome-White.svg';
```
Comment (thescientist13) [inline src/components/header/header.js]: @Peter-holyoke 
Let's keep all file names consistent, e.g. _slack-monochrome-white.svg_  (all lowercase)

File: src/images/Logo.svg [hunk 17]
```diff
@@ -0,0 +1,71 @@
+<?xml version="1.0" standalone="no"?>
```
Comment (thescientist13) [inline src/images/Logo.svg]: it seems that it would make sense to keep the logo SVG with the logo component itself.  Ideally, with components, everything it needs should ideally be co-located

**Comment by ChrisPlease** [comment]
@thescientist13 bootstrap's dist css won't have any mixins/variables available to access. The dist css has to be imported globally into the app regardless, so any component has access to bootstrap's classnames  (maybe there's a way to work with the raw pre-processed files, but Bootstrap 3 uses LESS anyway). IMHO, best practice would be to import the dist css at the entry file.

Additionally, [Bootstrap 3's module doesn't export anything](https://getbootstrap.com/docs/3.3/getting-started/#download-npm), so there's really no reason to import it [here](https://github.com/ProvidenceGeeks/website-frontend/blob/master/src/vendor.js#L1). 

I _do_ like the idea of keeping global scss file(s) (personally love Sass partials) that store things like global sass variables, defaults, global settings, etc. Component stylesheets can import one or more partials as needed, but I think naming it bootstrap might be confusing to new developers.

File: src/components/header/header.js (hunk 16 above)
Comment (Peter-holyoke) [reply src/components/header/header.js]: Will do. Also, I think it makes sense to use all SVGs for the social Icons. I'll discuss this with Elana, since it looks like it will require a change to one of the icons.

**Comment by thescientist13** [comment]
@ChrisPlease 
Good point about us importing the CSS and thus it not being shareable.  I guess for me, it seems more familiar / intuitive that any CSS `import` ing would happen within CSS files, with the exception being the CSS file of the component itself.  To me that seems like a more natural developer experience, but maybe I'm just alone in that feeling 😆 

You did point out something that I hadn't realized, in that we aren't using Bootstrap 4, so I will a ticket for that.  It will probably be a breaking change to this layout though : /  ( @Peter-holyoke )

Thanks so much for the review and great feedback!
